File: face_comparsion/src/main/java/netposa/realtime/kafka/utils/DirectKafkaConsumer.scala
package netposa.realtime.kafka.utils

import java.util
import java.util.concurrent.TimeUnit
import java.util.{Collections, Properties}

import kafka.common.TopicAndPartition
import kafka.utils.Logging
import org.apache.kafka.clients.consumer.{KafkaConsumer, OffsetAndMetadata, OffsetCommitCallback}
import org.apache.kafka.common.TopicPartition

import scala.collection.{JavaConversions, mutable}
import scala.collection.JavaConversions._
import scala.collection.mutable.{Buffer, ListBuffer}
import java.util.{List => JList}

class DirectKafkaConsumer(val props: Properties,val topic: String) extends Logging {

  val kafka =getConsumer

  private def getConsumer(): KafkaConsumer[Array[Byte], Array[Byte]] = {
    try {
      createConsumer()
    } catch {
      case e: Exception =>
        error(s"create kafka consumer failed! sleep 30s and try again. info is ${e.getLocalizedMessage}",e)
        TimeUnit.SECONDS.sleep(30.toLong)
        getConsumer()
    }
  }

  private def createConsumer():KafkaConsumer[Array[Byte],Array[Byte]]={
    val kafka_prop =new Properties()
    props.stringPropertyNames().foreach(key => {
      if (key.startsWith("kafka.")) {
        val key_p = key.stripPrefix("kafka.")
        val va = props.getProperty(key)
         key_p  match {
           case "auto.offset.reset"=>
              va.toLowerCase match {
                case "smallest" => kafka_prop.setProperty("auto.offset.reset", "earliest")
                case "largest" => kafka_prop.setProperty("auto.offset.reset", "latest")
                case _ => kafka_prop.put(key_p, va)
              }
           case "metadata.broker.list" =>kafka_prop.setProperty("bootstrap.servers", va)
           case _ =>   kafka_prop.put(key_p, va)
         }
      }
    })
    kafka_prop.setProperty("group.id",kafka_prop.getProperty(String.format("%s.group.id",topic)))
    //设置brokerServer(kafka)ip地址
    setDefultProps(kafka_prop,"enable.auto.commit", "false");
    // setDefultProps(props,"auto.offset.reset", "earliest")
    setDefultProps(kafka_prop,"key.deserializer","org.apache.kafka.common.serialization.ByteArrayDeserializer")
    setDefultProps(kafka_prop,"value.deserializer"," org.apache.kafka.common.serialization.ByteArrayDeserializer")
    setDefultProps(kafka_prop,"max.poll.records", "1000")
    println(kafka_prop)
    //topic msg
    val kafkaConsumer = new KafkaConsumer[Array[Byte],Array[Byte]](kafka_prop)
    kafkaConsumer.subscribe(JavaConversions.asJavaCollection(mutable.Buffer(kafka_prop.getProperty(topic + ".service.receive.topic.name"))))
    kafkaConsumer
  }


  def fetchMessages():Buffer[RowMessage]={
    val buffer = Buffer[RowMessage]()
    val allPartis =kafka.poll(props.getProperty("poll.timeout.ms","3000").toLong)
    allPartis.partitions().foreach(par=>{
      allPartis.records(par).foreach(v=>{
        buffer +=RowMessage(TopicAndPartition(par.topic(),par.partition()),v.value(),v.offset())
      })
    })
//    info(s"====This  batch  has fetched  Message  size is  [${buffer.size}]====")
    buffer
  }

  /**
    * 关闭消息接收服务
    */
  def shutdown() {
    kafka.close()
  }

  def commitOffsets(offsets: Map[TopicAndPartition, Long]):Unit={
    offsets.foreach(v=>{
      val offset= Collections.singletonMap(new TopicPartition(v._1.topic,v._1.partition),new OffsetAndMetadata(v._2+1))
      kafka.commitAsync(offset,new OffsetCommitCallback {
        override def onComplete(offsets: util.Map[TopicPartition, OffsetAndMetadata], exception: Exception) = {
          if(exception !=null) {
            error(s"commit offset failed! info is ${exception.getLocalizedMessage}")
          }
          debug(s"commit offset is ${offset}")
        }
      })
    })
  }

  private def setDefultProps(pro:Properties,key:String,value:String):Properties={
    if(!pro.containsKey(key)){
      pro.put(key,value)
    }
    props
  }

  private def initProps(): Map[String, String] = {
    val arr = new ListBuffer[(String, String)]()
    val propNames = {
      import JavaConversions._
      Collections.list(props.propertyNames).map(_.toString).sorted
    }
    propNames.foreach(key => {
      info("load properties key is [%s], value is [%s]".format(key, props.getProperty(key)))
      val row = (key, props.getProperty(key))
      arr += row
    })
      arr.toMap
  }
}
case class RowMessage(val tp:TopicAndPartition, val message:Array[Byte], val offset:Long) {
}

File: kafka_consumer_10/src/main/java/com/nukil/kafka/util/consumer/DirectKafkaConsumer.java
package com.nukil.kafka.util.consumer;

import com.nukil.kafka.util.bean.RowMessage;
import com.nukil.kafka.util.inter.ConsumerInterface;
import kafka.common.TopicAndPartition;
import org.apache.kafka.clients.consumer.*;
import org.apache.kafka.common.TopicPartition;
import org.apache.log4j.Logger;

import java.util.*;

/**
 * Created by Nukil on 2017/12/04
 */
public class DirectKafkaConsumer implements ConsumerInterface {
    private Properties properties;
    private boolean debug;
    private KafkaConsumer consumer;
    private Logger logger = Logger.getLogger(DirectKafkaConsumer.class);

    public DirectKafkaConsumer(Properties properties) {
        this.properties = properties;
        setDefaultConfig();
        this.debug = Boolean.parseBoolean(this.properties.getProperty("enable.kafka.consumer.debug", "false").trim());
        createConsumer();
    }

    @Override
    public void createConsumer() {
        try {
            consumer = new KafkaConsumer(properties);
            consumer.subscribe(Collections.singletonList(properties.getProperty("service.receive.topic.name")));
        } catch (Exception e) {
            logger.error("failed to create kafka consumer, sleep 30s and retry, error message is " + e.getMessage(), e);
            try {
                Thread.sleep(30000);
            } catch (Exception e1) {
                logger.error(e.getMessage(), e1);
            }
            createConsumer();
        }
    }

    @Override
    public List<RowMessage> fetchMessages() {
        List<RowMessage> messages = new ArrayList<>();
        ConsumerRecords<byte[], byte[]> records = consumer.poll(Long.parseLong(properties.getProperty("poll.timeout.ms", "3000")));
        for (ConsumerRecord<byte[], byte[]> record : records) {
            messages.add(new RowMessage(new TopicAndPartition(record.topic(), record.partition()), record.value(), record.offset()));
        }
        return messages;
    }

    @Override
    public boolean commitOffsets(Map<TopicAndPartition, Long> offsets) {
        for (TopicAndPartition key : offsets.keySet()) {
            final Map<TopicPartition, OffsetAndMetadata> offset = new HashMap<>();
            offset.put(new TopicPartition(key.topic(), key.partition()), new OffsetAndMetadata(offsets.get(key) + 1));
            consumer.commitAsync(offset, new OffsetCommitCallback() {
                @Override
                public void onComplete(Map<TopicPartition, OffsetAndMetadata> map, Exception e) {
                    if (null != e) {
                        logger.error(e.getLocalizedMessage(), e);
                    }
                    if (debug) {
                        logger.info("submit offset : " + offset);
                    }
                }
            });
        }
        return true;
    }

    @Override
    public void shutDown() {
        if (null != consumer) {
            logger.info("try to shutdown consumer......");
            consumer.close();
        }
    }

    @Override
    public void setDefaultConfig() {
        Set keySet = properties.keySet();
        if (!keySet.contains("enable.auto.commit")) {
            properties.setProperty("enable.auto.commit", "false");
        }
        if (!keySet.contains("value.deserializer")) {
            properties.setProperty("value.deserializer", "org.apache.kafka.common.serialization.ByteArrayDeserializer");
        }
        if (!keySet.contains("key.deserializer")) {
            properties.setProperty("key.deserializer", "org.apache.kafka.common.serialization.ByteArrayDeserializer");
        }
    }
}